### PR TITLE
C#: Data flow through fields

### PR DESCRIPTION
**Pull request by hvitved**
Initial implementation of data flow through fields, using the algorithm of the shared data flow implementation. Fields (and field-like properties) are covered, and stores can be either
 - ordinary assignments, `Foo = x`,
 - object initializers, `new C() { Foo = x }`, or
 - field initializers, `int Foo = x`.

Update: For field initializers, the CFG has been extended by adding the initializer expressions to the relevant constructors. For example, in

```
class C
{
    int Field1 = 0;
    int Field2 = Field1 + 1;
    int Field3;

    public C()
    {
        Field3 = 2;
    }

    public C(int i)
    {
        Field3 = 3;
    }
}
```

the initializer expressions `Field1 = 0` and `Field2 = Field1 + 1` are added to the two constructors, thus mimicking

```
public C()
{
    Field1 = 0;
    Field2 = Field1 + 1;
    Field3 = 2;
}
```

and

```
public C()
{
    Field1 = 0;
    Field2 = Field1 + 1;
    Field3 = 3;
}
```

respectively.

~~For field initializers, we need to synthesize calls (`SynthesizedCall`), callables (`SynthesizedCallable`), parameters (`InstanceParameterNode`), and arguments (`SynthesizedThisArgumentNode`), as the C# extractor does not (yet) extract such entities. For example, in~~

~~class C
{
    int Field1 = 1;
    int Field2 = 2;
    C() { }
}~~

~~there is a synthesized call from the constructor `C`, with a synthesized `this` argument, and the targets of that call are two synthesized callables with bodies `this.Field1 = 1` and `this.Field2 = 2`, respectively.~~

~~A consequence of this is that `DataFlowCallable` is no longer an alias for `DotNet::Callable`, but instead an IPA type.~~

----

For review of the test `FieldFlow.ql`, I recommend running the query in QL4E and explore the data flow paths using the path explorer.

Future work includes:
- Type pruning (as in the Java library).
- Flow through arrays/collections.
- Flow summaries for library types take access paths into account, for example a proper model for `System.Tuple`s.
- Model flow through captured variables as flow through fields.

@calumgrant (also ping @aschackmull)

### CODE REVIEW

**Comment by hvitved** [comment]
I ran a [dist-compare](https://git.semmle.com/gist/tom/2037c5e7179354079db237251989ca62#file-report-md) (internal link), which revealed a significant slow-down, particular on mono, where we timeout on `InappropriateEncoding.ql`. This is partly due to overlapping data flow configurations (addressed by 1e46509a2afff71b57065af088efd94dfa772390), partly due to bad join-orders related to field flow (addressed by a0c834c83d7b707451ef9ae22ac05fbb2653a3f9, ping @aschackmull ). I have started a new dist-compare run.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll [hunk 1]
```diff
@@ -257,9 +257,8 @@ private predicate nodeCandFwd1(Node node, boolean stored, Configuration config)
     )
     or
     // read
-    exists(Node mid, Content f |
-      nodeCandFwd1(mid, true, config) and
-      read(mid, f, node) and
+    exists(Content f |
+      nodeCandFwd1Read(f, node, config) and
       storeCandFwd1(f, unbind(config)) and
```
Comment (aschackmull) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll]: With this predicate factoring we should be able to remove the `unbind`.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll [hunk 2]
```diff
@@ -337,10 +345,9 @@ private predicate nodeCand1(Node node, boolean stored, Configuration config) {
     )
     or
     // store
-    exists(Node mid, Content f |
-      store(node, f, mid) and
+    exists(Content f |
+      nodeCand1Store(f, node, config) and
       readCand1(f, unbind(config)) and
```
Comment (aschackmull) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll]: Similarly here, the `unbind` shouldn't be necessary, as the optimiser ought to then do a direct multi-column join on the two recursive calls.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll [hunk 3]
```diff
@@ -648,9 +665,8 @@ private predicate nodeCandFwd2(Node node, boolean fromArg, boolean stored, Confi
     )
     or
     // read
-    exists(Node mid, Content f |
-      nodeCandFwd2(mid, fromArg, true, config) and
-      read(mid, f, node) and
+    exists(Content f |
+      nodeCandFwd2Read(f, node, fromArg, config) and
       storeCandFwd2(f, unbind(config)) and
```
Comment (aschackmull) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll]: Same.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll [hunk 4]
```diff
@@ -721,10 +747,9 @@ private predicate nodeCand2(Node node, boolean toReturn, boolean stored, Configu
     )
     or
     // store
-    exists(Node mid, Content f |
-      store(node, f, mid) and
+    exists(Content f |
+      nodeCand2Store(f, node, toReturn, config) and
       readCand2(f, unbind(config)) and
```
Comment (aschackmull) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll]: Same.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll [hunk 5]
```diff
@@ -765,25 +791,30 @@ private predicate readCand2(Content f, Configuration config) {
...
+
 pragma[nomagic]
 private predicate storeCand(Content f, Configuration conf) {
-  exists(Node n1, Node n2 |
-    store(n1, f, n2) and
-    nodeCand2(n1, _, _, conf) and
-    nodeCand2(n2, _, _, unbind(conf))
+  exists(Node node |
+    nodeCand2Store(f, node, _, conf) and
+    nodeCand2(node, _, _, unbind(conf))
```
Comment (aschackmull) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll]: Same.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll [hunk 6]
```diff
@@ -1028,6 +1058,17 @@ private predicate consCandFwd(Content f, AccessPathFront apf, Configuration conf
   )
 }
 
+pragma[noinline]
+private predicate flowCandFwdRead(
+  Content f, Node node, AccessPathFront apf, boolean fromArg, Configuration config
```
Comment (aschackmull) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll]: Could you switch the order of `apf` and `fromArg` to match the order in `flowCandFwd`?

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll [hunk 7]
```diff
@@ -1041,9 +1041,9 @@ private predicate flowCandFwd0(Node node, boolean fromArg, AccessPathFront apf,
   )
   or
   exists(Content f, AccessPathFront apf0 |
-    flowCandFwdRead(f, node, apf0, fromArg, config) and
+    flowCandFwdRead(f, node, fromArg, apf0, config) and
     apf0.headUsesContent(f) and
```
Comment (aschackmull) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll]: I think this line ought to be pushed into `flowCandFwdRead` as well.

Comment (aschackmull) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll]: (and then the `apf` argument can be projected away)

Comment (hvitved) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll]: Makes sense.

**Comment by calumgrant** [comment]
I'm struggling a little bit with the idea that member initializers are "callables", and that the initializers are invoked by calling them. There is a lot of additional machinery required to support this, and I wonder if this abstraction is temporary until we model initializers differently anyway. Therefore, I wonder whether the CFG of the constructor should instead create CFG nodes relating to the field initializers. Do you think this is an idea that is worth prototyping?

**Comment by hvitved** [comment]
> Therefore, I wonder whether the CFG of the constructor should instead create CFG nodes relating to the field initializers. Do you think this is an idea that is worth prototyping?

That might actually be a good idea; I will try it out. Thanks!

**Comment by aschackmull** [comment]
Consider
```
class A {
  int f1 = 1;
  int f2 = 2;
  A() { }
  A(int x) { }
}
```
If you insert CFG nodes for the field initializers in the constructors then you'll have multiple enclosing callables for them, which is slightly awkward.  In Java the extractor synthesizes a single callable called `<obinit>`, which contains all the field inits, so the above becomes:
```
class A {
  int f1;
  int f2;
  <obinit>(A this) {
    this.f1 = 1;
    this.f2 = 2;
  }
  A() {
    <obinit>(this);
  }
  A(int x) {
    <obinit>(this);
  }
}
```

**Comment by hvitved** [comment]
> If you insert CFG nodes for the field initializers in the constructors then you'll have multiple enclosing callables for them

My initial idea was to use splitting, i.e. have one copy of each assignment per constructor.

**Comment by hvitved** [comment]
Here is the updated [dist-compare](https://git.semmle.com/gist/tom/2037c5e7179354079db237251989ca62#file-report2-md) report (internal link), up until 7ab9c8b90d6fd80c7b9c69f159445b1f8ee52cad. It shows that the performance optimizations have paid off, but mono is still considerably slower than without field flow.

**Comment by hvitved** [comment]
@calumgrant : I have update this PR with the change that you suggested. I think that works much better, and it has the advantage that is works for all CFG-based analysis, not just data flow. I have started a new `dist-compare` run.

**Comment by hvitved** [comment]
For completeness, [here](https://git.semmle.com/gist/tom/2037c5e7179354079db237251989ca62#file-report4-md) is the final dist-compare report.

**Pull request merged**